Fix testsuite --chain-lint option, add timeout to some tests, and other testing fixes

grondo opened the pull request:
This branch started out as a small fix to propagate sharness' `--chain-lint` option to `test_under_flux` so that broken `&&`-chains can be detected for tests that run under a flux instance.

After that, I went through all the tests and fixed broken `&&`-chains so that `--chain-lint` ran clean.
In most cases the broken `&&` chain wasn't an actual problem, but I fixed the test anyway. In other cases it wasn't possible to make `--chain-lint` happy (e.g. test that use `&` to run a process in the background). For these cases I added a `NO_CHAIN_LINT` prereq that can be explicitly added to a tests so that test is skipped during a run with `--chain-lint`.

Finally, the `flux version` check (once it was fixed) didn't work in travis (at least on my personal branch) because travis uses a shallow clone of the repo. A workaround was added to `.travis.yml` to fetch tags and "unshallow" the clone so that hopefully `git describe` can create a valid version.

grondo commented:
This PR has been having trouble with intermittent hangs in Travis. Not sure what is going on here, but another variable is that Travis-CI is transitioning to their Trusty environment (Ubuntu 14.04). However, other PRs don't seem to have a problem, so this one's not a merge candidate for now.

garlick commented:
Just had a spin through and looks like a lot of good changes.  I didn't spot anything wrong.

grondo commented:
Still can't pin down why this PR seems to cause intermittent but frequent hangs in Travis.

What's the equivalent of tearing a PR into pieces, throwing in the garbage, and lighting it on fire?

grondo commented:
Ok, I reworked the changes in t1002-kvs-cmd.t and now the hangs seem to have stopped, though I didn't see anything wrong that would cause hangs. I suppose also something could be different this AM in Travis itself. :man_shrugging: 

I also added to this PR some debug for issue #997 (will fix the typo when we figure the problem with the test)

grondo commented:
I spoke too soon, one of the tests (for grondo/flux-core, not flux-framework/flux-core) did hang, but very early in the unit tests (apparently, difficult to tell) before the sharness tests under `t` even started:
```
PASS: test_tagpool.t 16 - group: tagpool_alloc worked 256 times
PASS: test_tagpool.t 17 - group: pool depleted by 256
PASS: test_tagpool.t 18 - group: allocated tags contain no duplicates
PASS: test_tagpool.t 19 - group: tagpool_free restored all to pool
PASS: test_tagpool.t 20 - group: entire poool allocated by tagpool_alloc loop
PASS: test_tagpool.t 21 - group: pool is exhausted
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

garlick commented:
21 is the last tagpool test.  Maybe the hanging one is libflux/test/security.c, which comes after (normally).  I wonder if it would make sense to call `alarm()` in that one, e.g.
```diff
diff --git a/src/common/libflux/test/security.c b/src/common/libflux/test/security.c
index 8b2bcaf..ede851f 100644
--- a/src/common/libflux/test/security.c
+++ b/src/common/libflux/test/security.c
@@ -469,10 +469,19 @@ void test_curve (void)
     unlink_recursive (path);
 }
 
+void alarm_callback (int arg)
+{
+    diag ("test timed out");
+    exit (1);
+}
+
 int main (int argc, char *argv[])
 {
     plan (NO_PLAN);
 
+    signal (SIGALRM, alarm_callback);
+    alarm (30);
+
     test_ctor_dtor ();
     test_keygen ();
     test_munge ();
```

grondo commented:
I rebased on master with @garlick's rc2_timeout patch and added a default timeout of 5m to all tests.

Unfortunately I don't think the timer is compatible with the libfaketime-based cron tests, as these are now always timing out, though I thought ev_timer watchers measured real time not wall clock time, which should be the only thing libfaketime affects (but I can't remember the exact distinctions right now).

For now, I'll add an option to disable the timeout on a per-test basis.

I'll also add @garlick's patch for test_security.t above because the travis runs hit a hang there as well.

grondo commented:
I pared this PR down to a very small set of fixes that do not seem to cause travis-ci hangs (today at least) anymore.

The only fixes now included is the fix to propagate `--chain-lint` through `test_under_flux`, addition of a default 5m timeout to all `test_under_flux` runs using the new `init.rc2_timeout` broker attr,  the addition of a timeout to test_security.t as proposed by @garlick, along with a couple other debug and codecov.io fixes.

I propose we just merge this for now, and then I can push test fixes on top one by one to see what the hangs were caused by (or just drop those fixes for now)

garlick commented:
Right let's do it!
